packages/block-editor/src/components/inserter/hooks/use-block-types-state.js

```diff
@@ -23,6 +23,25 @@ import { store as blockEditorStore } from '../../../store';
  * @return {Array} Returns the block types state. (block types, categories, collections, onSelect handler)
  */
 const useBlockTypesState = ( rootClientId, onInsert ) => {
+	const [ allItems ] = useSelect( ( select ) => {
+		let availableItems = select( blockEditorStore ).getInserterItems( '' );
+
+		// use current selection as root for situations like
+		// template locked mode
+		const rootBlocks = select( blockEditorStore ).getBlocks();
+		while ( availableItems.length === 0 ) {
+			for ( const block of rootBlocks ) {
```

Comment:
This loop feels like it can be very bad perf wise in nested templates and things like that. (not sure how true though).

The other thing that feels odd, is that it will end up with a random client id and the behavior for the user can be confusing: where am I going to insert when I click that thing?

It feels like a better path, would be to try to find the "rootClientId" that we want to insert in before hand and update `useBlockTypesState` to take a rootClientId as an argument. Now that I'm writing that, it seems it already has a rootClientId, so it's a matter of extracting a "BlockList" component from the current BlockTab component and doing something like:

```
const { rootClientId, alternativeRootClientId } = useSelect(() => {
   return {
       rootClientId: getInsertionPoint().rootClientId,
       alternativeRootClientId: findAlternativeRootClientId() // we can discuss implementation of this function separately.
   }
}, [] );

<BlockList rootClientId={ rootClientID } />
<BlockList rootClientId={ alternativeRootClientId } />
```

Reply:
Yes by all means there a couple of risky loops but they're in place just to make things "work" and to show conceptual limitations we still have.

For example a loop later in the PR will make the action of inserting a block while a post content block is selected in a template locked edit mode and it shows that we can't tell where to insert without:

- bringing knowledge of template locked mode to the block editor package
- relying on something else in a higher level package finding "post content" and its client id and pass that down to lower levels

... both of which are kinda weird.

Separately, the theoretical `findAlternativeRootClientId` will also have to try a sort of heuristic since we don't have a clear picture of how to find the next best insertable spot.

Maybe the block list should advertise this in a separate precomputed part of the store? As in top down places to insert blocks if the next thing is unavailable ... or something.

Reply:
Yeah, sorry I misread the original reply :)